### PR TITLE
Render source page layout with Askama

### PR DESCRIPTION
**Pull request by clubby789**
~~I was looking at making `code_html` render into the buffer instead of in advance, but it turned out to need a pretty big refactor, so starting with rearranging the high-level layout.~~
Found another approach which required much less changes

cc #108868

### CODE REVIEW

**Review by notriddle — CHANGES_REQUESTED** [review]
Please do not silently discard errors without a comment explaining why.

File: src/librustdoc/html/templates/source.html [hunk 1]
```diff
@@ -0,0 +1,19 @@
+<div class="example-wrap"> {# #}
+    <pre class="src-line-numbers">
+        {% for line in lines.clone() %}
+            {% if embedded %}
+                <span>{{line}}</span>
+            {%~ else %}
+                <a href="#{{line}}" id="{{line}}">{{line}}</a>
+            {%~ endif %}
+        {% endfor %}
+    </pre>
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/templates/source.html:10]: Missing a `{# #}` here.

**Review by GuillaumeGomez — COMMENTED** [review]
Other than my 2 comments, looks good to me. :+1:

File: src/librustdoc/html/templates/source.html [hunk 2]
```diff
@@ -0,0 +1,19 @@
+<div class="example-wrap"> {# #}
+    <pre class="src-line-numbers">
+        {% for line in lines.clone() %}
```
Comment (klensy) [inline src/librustdoc/html/templates/source.html:3]: lines is `Range`, so it's free, but is clone required here?

Comment (clubby789) [reply src/librustdoc/html/templates/source.html:3]: iirc, without that Askama tries to move it

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
We have some small regressions but the code improvement is worth it I think.

What do you think @notriddle ?

**Comment by GuillaumeGomez** [comment]
Well, that's definitely worth a try.

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Seems like the regression was fixed, well done!

@bors r+ rollup

**Pull request merged**